{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pandas as pd\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_file = \"/data5/deepro/ukbiobank/papers/bmi_project/0_data_download/published_studies/GWAS_Catalog/data/gwas_genes_to_traits.csv\"\n",
    "combinations2_file = \"/data5/deepro/ukbiobank/papers/bmi_project/3_run_rarecomb/white_british/data/parsed_tables/combo_2.csv\"\n",
    "combinations3_file = \"/data5/deepro/ukbiobank/papers/bmi_project/3_run_rarecomb/white_british/data/parsed_tables/combo_3.csv\"\n",
    "gencode_file = \"/data5/deepro/ukbiobank/papers/bmi_project/1_parse_data/prepare_gencode_genes/data/gencode.v39.parsed.genes.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_df = pd.read_csv(gwas_file, low_memory=False)\n",
    "combinations2_df = pd.read_csv(combinations2_file, low_memory=False)\n",
    "combinations3_df = pd.read_csv(combinations3_file, low_memory=False)\n",
    "gencode_df = pd.read_csv(gencode_file).drop_duplicates('gene_id_stripped').set_index('gene_id_stripped', drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get genes being studied\n",
    "genes = pd.read_csv(\"/data5/deepro/ukbiobank/papers/bmi_project/1_parse_data/annotate_vcf/data/variants_by_gene/lof_missense_pred_freq_0.01_format2.tsv\", sep='\\t', nrows=0)\n",
    "genes = list(genes.columns)[1:]\n",
    "genes = [s.split('_')[1] for s in genes]\n",
    "genes = list(gencode_df.loc[genes]['gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gwas_enrichment_table(combo_dfs, ncombos_list, gwas_df, genes_study, save_file):\n",
    "    # restrict to genes present both in GWAS and all analyzed in our study\n",
    "    genes_study = list(set(genes_study).intersection(set(gwas_df.Gene)))\n",
    "    gwas_df = gwas_df.loc[gwas_df.Gene.isin(genes_study)]\n",
    "    gwas_genes = set(gwas_df.Gene)\n",
    "    # get list of unique genes in combos\n",
    "    unique_combo_genes = set()\n",
    "    for combo_df, ncombo in zip(combo_dfs, ncombos_list):\n",
    "        combo_set = set(combo_df.loc[:, [f\"Item_{i}_symbol\" for i in range(1, ncombo + 1)]].values.flatten())\n",
    "        unique_combo_genes.update(combo_set)\n",
    "    # keep only those present in gwas database\n",
    "    unique_combo_genes = unique_combo_genes.intersection(gwas_genes)\n",
    "    stats = []\n",
    "    # only keep phenotypes with at least 10 genes\n",
    "    trait_counts = gwas_df.Mapped_Trait.value_counts()\n",
    "    trait_keep = list(trait_counts.loc[trait_counts >= 10].index)\n",
    "    gwas_df = gwas_df[gwas_df.Mapped_Trait.isin(trait_keep)]\n",
    "    for trait in gwas_df.Mapped_Trait.unique():\n",
    "        all_genes_with_trait = set(gwas_df.loc[gwas_df.Mapped_Trait==trait, \"Gene\"])\n",
    "        all_genes_without_trait = gwas_genes.difference(all_genes_with_trait)\n",
    "        combo_genes_with_trait = unique_combo_genes.intersection(all_genes_with_trait)\n",
    "        combo_genes_without_trait = unique_combo_genes.difference(combo_genes_with_trait)\n",
    "        noncombo_genes_with_trait = all_genes_with_trait.difference(unique_combo_genes)\n",
    "        noncombo_genes_without_trait = all_genes_without_trait.difference(unique_combo_genes)\n",
    "\n",
    "        contingency_table = [[len(combo_genes_with_trait), len(combo_genes_without_trait)],\n",
    "                            [len(noncombo_genes_with_trait), len(noncombo_genes_without_trait)]]\n",
    "        \n",
    "        result = fisher_exact(contingency_table, alternative='greater')\n",
    "        oddsratio = result[0]\n",
    "        pvalue = result[1]\n",
    "        stats.append([trait, oddsratio, pvalue, len(combo_genes_with_trait), len(combo_genes_without_trait), len(noncombo_genes_with_trait), len(noncombo_genes_without_trait)])\n",
    "    # multiple testing for stats df\n",
    "    stats = pd.DataFrame(stats, columns=['gwas_phenotype', 'oddsratio', 'pvalue', 'Num_combo_genes_with_phenotype', 'Num_combo_genes_without_phenotype', 'Num_noncombo_genes_with_phenotype', 'Num_noncombo_genes_without_phenotype'])\n",
    "    stats['FDR'] = multipletests(stats.pvalue, method='fdr_bh')[1]\n",
    "    # save file\n",
    "    stats.sort_values(\"FDR\").to_csv(save_file, index=False)\n",
    "    return stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_file = \"/data5/deepro/ukbiobank/papers/bmi_project/4_characterization/white_british/data/enrichment/gwas_enrichment.csv\"\n",
    "stats = create_gwas_enrichment_table([combinations2_df, combinations3_df], [2, 3], gwas_df, genes, save_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gwas_phenotype</th>\n",
       "      <th>oddsratio</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Num_combo_genes_with_phenotype</th>\n",
       "      <th>Num_combo_genes_without_phenotype</th>\n",
       "      <th>Num_noncombo_genes_with_phenotype</th>\n",
       "      <th>Num_noncombo_genes_without_phenotype</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>body height</td>\n",
       "      <td>2.187713</td>\n",
       "      <td>1.969847e-04</td>\n",
       "      <td>32</td>\n",
       "      <td>127</td>\n",
       "      <td>1496</td>\n",
       "      <td>12989</td>\n",
       "      <td>0.033226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>intraocular pressure measurement</td>\n",
       "      <td>3.501546</td>\n",
       "      <td>3.584181e-04</td>\n",
       "      <td>12</td>\n",
       "      <td>147</td>\n",
       "      <td>330</td>\n",
       "      <td>14155</td>\n",
       "      <td>0.035692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>urate measurement</td>\n",
       "      <td>2.803782</td>\n",
       "      <td>4.317373e-05</td>\n",
       "      <td>23</td>\n",
       "      <td>136</td>\n",
       "      <td>824</td>\n",
       "      <td>13661</td>\n",
       "      <td>0.010319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>antibody measurement</td>\n",
       "      <td>27.836538</td>\n",
       "      <td>3.316356e-04</td>\n",
       "      <td>3</td>\n",
       "      <td>156</td>\n",
       "      <td>10</td>\n",
       "      <td>14475</td>\n",
       "      <td>0.035692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>gut microbiome measurement</td>\n",
       "      <td>4.135730</td>\n",
       "      <td>6.098362e-07</td>\n",
       "      <td>20</td>\n",
       "      <td>139</td>\n",
       "      <td>487</td>\n",
       "      <td>13998</td>\n",
       "      <td>0.000729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>bone density</td>\n",
       "      <td>3.059050</td>\n",
       "      <td>3.040483e-04</td>\n",
       "      <td>15</td>\n",
       "      <td>144</td>\n",
       "      <td>477</td>\n",
       "      <td>14008</td>\n",
       "      <td>0.035692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>periodontitis</td>\n",
       "      <td>6.330441</td>\n",
       "      <td>2.878884e-05</td>\n",
       "      <td>9</td>\n",
       "      <td>150</td>\n",
       "      <td>136</td>\n",
       "      <td>14349</td>\n",
       "      <td>0.010319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>executive function measurement</td>\n",
       "      <td>27.836538</td>\n",
       "      <td>3.316356e-04</td>\n",
       "      <td>3</td>\n",
       "      <td>156</td>\n",
       "      <td>10</td>\n",
       "      <td>14475</td>\n",
       "      <td>0.035692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>Epstein-Barr virus infection</td>\n",
       "      <td>10.191276</td>\n",
       "      <td>2.224329e-04</td>\n",
       "      <td>5</td>\n",
       "      <td>154</td>\n",
       "      <td>46</td>\n",
       "      <td>14439</td>\n",
       "      <td>0.033226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>826</th>\n",
       "      <td>BMI-adjusted waist circumference</td>\n",
       "      <td>3.162185</td>\n",
       "      <td>4.986423e-06</td>\n",
       "      <td>24</td>\n",
       "      <td>135</td>\n",
       "      <td>771</td>\n",
       "      <td>13714</td>\n",
       "      <td>0.002979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>893</th>\n",
       "      <td>BMI-adjusted hip circumference</td>\n",
       "      <td>2.591159</td>\n",
       "      <td>9.650981e-05</td>\n",
       "      <td>24</td>\n",
       "      <td>135</td>\n",
       "      <td>930</td>\n",
       "      <td>13555</td>\n",
       "      <td>0.019222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1069</th>\n",
       "      <td>appendicular lean mass</td>\n",
       "      <td>3.052699</td>\n",
       "      <td>4.050080e-05</td>\n",
       "      <td>20</td>\n",
       "      <td>139</td>\n",
       "      <td>652</td>\n",
       "      <td>13833</td>\n",
       "      <td>0.010319</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gwas_phenotype  oddsratio        pvalue  \\\n",
       "3                          body height   2.187713  1.969847e-04   \n",
       "57    intraocular pressure measurement   3.501546  3.584181e-04   \n",
       "70                   urate measurement   2.803782  4.317373e-05   \n",
       "165               antibody measurement  27.836538  3.316356e-04   \n",
       "167         gut microbiome measurement   4.135730  6.098362e-07   \n",
       "286                       bone density   3.059050  3.040483e-04   \n",
       "356                      periodontitis   6.330441  2.878884e-05   \n",
       "397     executive function measurement  27.836538  3.316356e-04   \n",
       "769       Epstein-Barr virus infection  10.191276  2.224329e-04   \n",
       "826   BMI-adjusted waist circumference   3.162185  4.986423e-06   \n",
       "893     BMI-adjusted hip circumference   2.591159  9.650981e-05   \n",
       "1069            appendicular lean mass   3.052699  4.050080e-05   \n",
       "\n",
       "      Num_combo_genes_with_phenotype  Num_combo_genes_without_phenotype  \\\n",
       "3                                 32                                127   \n",
       "57                                12                                147   \n",
       "70                                23                                136   \n",
       "165                                3                                156   \n",
       "167                               20                                139   \n",
       "286                               15                                144   \n",
       "356                                9                                150   \n",
       "397                                3                                156   \n",
       "769                                5                                154   \n",
       "826                               24                                135   \n",
       "893                               24                                135   \n",
       "1069                              20                                139   \n",
       "\n",
       "      Num_noncombo_genes_with_phenotype  Num_noncombo_genes_without_phenotype  \\\n",
       "3                                  1496                                 12989   \n",
       "57                                  330                                 14155   \n",
       "70                                  824                                 13661   \n",
       "165                                  10                                 14475   \n",
       "167                                 487                                 13998   \n",
       "286                                 477                                 14008   \n",
       "356                                 136                                 14349   \n",
       "397                                  10                                 14475   \n",
       "769                                  46                                 14439   \n",
       "826                                 771                                 13714   \n",
       "893                                 930                                 13555   \n",
       "1069                                652                                 13833   \n",
       "\n",
       "           FDR  \n",
       "3     0.033226  \n",
       "57    0.035692  \n",
       "70    0.010319  \n",
       "165   0.035692  \n",
       "167   0.000729  \n",
       "286   0.035692  \n",
       "356   0.010319  \n",
       "397   0.035692  \n",
       "769   0.033226  \n",
       "826   0.002979  \n",
       "893   0.019222  \n",
       "1069  0.010319  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.loc[stats.FDR<0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "starrseq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "06856c2527dea3cf472b3c4994ec564df76496982852535c5a9520cd26601212"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
